src/43F0.c

```diff
@@ -74,9 +74,128 @@ HeapNode* _heap_create(s32* addr, u32 size) {
     }
 }
 
-INCLUDE_ASM(s32, "43F0", _heap_malloc);
+#define _heap_alloc_and_update_id(node) { \
+    u16 HeapEntryID = D_NEXT_HEAP_MALLOC_ID; \
+    node->allocated = 1; \
+    D_NEXT_HEAP_MALLOC_ID = HeapEntryID + 1; \
+    node->entryID = HeapEntryID; \
+}
+
+void* _heap_malloc(HeapNode* head, u32 size) {
+    HeapNode* nextHeapNode;
+    HeapNode* pPrevHeapNode = 0;
+    u32 newBlockSize;
+    u32 curBlockLength;
+    HeapNode* curHeapNode;
+    u32 smallestBlockFound;
+
+    //must allocate 16 bytes or more at minimum or fail
+    size = ALIGN16(size);
+    if (size == 0) {
+        return (s32 *)pPrevHeapNode;
+    }
+
+    smallestBlockFound = 0;
+    nextHeapNode = 0;
+
+    //find the smallest block we can fit into in the free list
+    for (curHeapNode = head;;curHeapNode = curHeapNode->next) {
+        if (curHeapNode->allocated == 0) {
+            curBlockLength = curHeapNode->length;
+            if ((curBlockLength >= size) && ((curBlockLength < smallestBlockFound) || (smallestBlockFound == 0))) {
+                pPrevHeapNode = curHeapNode;
+                smallestBlockFound = curBlockLength;
+                nextHeapNode = curHeapNode->next;
+            }
+        }
+        if (!curHeapNode->next) {
+            break;
+        }
+    }
+
+
+    //find out the required block size with header
+    newBlockSize = size + sizeof(HeapNode);
+
+    //if we found a block see if we need to split it up
+    if (smallestBlockFound) {
+        if (smallestBlockFound >= newBlockSize) {
+            //update previous to the proper size for the block being returned
+            pPrevHeapNode->next = (HeapNode *)((u8 *)pPrevHeapNode + newBlockSize);
+            pPrevHeapNode->length = size;
+            _heap_alloc_and_update_id(pPrevHeapNode);
+
+            //setup the new heap block entry
+            curHeapNode = pPrevHeapNode->next;
+            curHeapNode->next = nextHeapNode;
+            curHeapNode->length = smallestBlockFound - newBlockSize;
+            curHeapNode->allocated = 0;
+        } else {
+            //take this entry out of the free linked list and mark as allocated
+            pPrevHeapNode->next = nextHeapNode;
+            pPrevHeapNode->length = smallestBlockFound;
+
+            //update the entry id on allocation
+            _heap_alloc_and_update_id(pPrevHeapNode);
+        }
+        return (u8 *)pPrevHeapNode + sizeof(HeapNode);
+    }
+    return NULL;
+}
 
-INCLUDE_ASM(s32, "43F0", _heap_malloc_tail);
+void* _heap_malloc_tail(HeapNode* head, u32 size) {
+    HeapNode *curNode;
+    u32 newNodeSize;
+    u32 foundNodeLength;
+    HeapNode *foundNode;
+    HeapNode *nextNode;
```

Comment:
We left-align pointers:
```suggestion
    HeapNode* curNode;
    u32 newNodeSize;
    u32 foundNodeLength;
    HeapNode* foundNode;
    HeapNode* nextNode;
```